arrow/src/datatypes/schema.rs

```diff
@@ -87,6 +87,18 @@ impl Schema {
         Self { fields, metadata }
     }
 
+
+    /// Returns a new schema with only the specified columns in the new schema
+    /// This carries metadata from the parent schema over as well
+    pub fn project(&self, indices: impl IntoIterator<Item=usize>) -> Result<Schema> {
+        let mut new_fields = vec![];
+        for i in indices {
+            let f = self.fields[i].clone();
+            new_fields.push(f);
+        }
```

Comment:
I think as written 
1.  This will `panic!` if there the index is not in bounds:
2.  is not "idiomatic rust style" (which to me means avoid `mut`). This is far less important

How about something such as (untested):

```suggestion
        let new_fields = indices
          .into_iter()
          .map(|i| {
            self.fields.get(i).map(|f| f.clone()))
              .ok_or_else(|| Err(ArrowError::SchemaError(
                format!("project index {} out of bounds, max field {}"
                                    i, self.fields().len()),
                            ))
          })
          .collect::<Result<Vec<_>>>()?;
```

Note the use of https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get to avoid `fields[i]` and then the somewhat confusing use of turbofish `.collect::<Result<Vec<_>>()` -- it took me quite a while to get used to that pattern

Reply:
Yea, that seems good to me, the for loop was the first thing that popped into my head, but I can't think of any reason it's better than yours.

Reply:
I think the for loop thing is what one would write in other languages like C/C++, Java, go ,etc :) It is certainly what I was writing when I started learning rust.

Then I realized that a big part of how rust avoids bounds checks while still being safe is by the use of the functional style (as in the compiler sets up the bound for `indicies` correctly once in the `into_iter()` and then doesn't need to do any checking on each iteration)